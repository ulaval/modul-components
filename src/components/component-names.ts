export const ACCORDION_GROUP_NAME: string = 'm-accordion-group';
export const ACCORDION_NAME: string = 'm-accordion';
export const LIST_ITEM_NAME: string = 'm-list-item';
export const BUTTON_NAME: string = 'm-button';
export const BUTTON_GROUP_NAME: string = 'm-button-group';
export const CAROUSEL_NAME: string = 'm-carousel';
export const CAROUSEL_ITEM_NAME: string = 'm-carousel-item';
export const CHECKBOX_NAME: string = 'm-checkbox';
export const DATEFIELDS_NAME: string = 'm-datefields';
export const DATEPICKER_NAME: string = 'm-datepicker';
export const DIALOG_NAME: string = 'm-dialog';
export const DROPDOWN_NAME: string = 'm-dropdown';
export const DROPDOWN_ITEM_NAME: string = 'm-dropdown-item';
export const DROPDOWN_GROUP_NAME: string = 'm-dropdown-group';
export const DYNAMIC_TEMPLATE_NAME: string = 'm-dynamic-template';
export const FILE_SELECT_NAME: string = 'm-file-select';
export const FILE_UPLOAD_NAME: string = 'm-file-upload';
export const EDIT_WINDOW_NAME: string = 'm-edit-window';
export const FLEX_TEMPLATE_NAME: string = 'm-flex-template';
export const I18N_NAME: string = 'm-i18n';
export const ICON_NAME: string = 'm-icon';
export const ICON_BUTTON_NAME: string = 'm-icon-button';
export const INPUT_STYLE_NAME: string = 'm-input-style';
export const LIMIT_TEXT_NAME: string = 'm-limit-text';
export const LINK_NAME: string = 'm-link';
export const MESSAGE_NAME: string = 'm-message';
export const MODAL_NAME: string = 'm-modal';
export const NAVBAR_NAME: string = 'm-navbar';
export const NAVBAR_ITEM_NAME: string = 'm-navbar-item';
export const MENU_NAME: string = 'm-menu';
export const MENU_ITEM_NAME: string = 'm-menu-item';
export const PANEL_NAME: string = 'm-panel';
export const PHONE_NUMBER_NAME: string = 'm-phone-number';
export const POPPER_NAME: string = 'm-popper';
export const POPUP_NAME: string = 'm-popup';
export const PROGRESS_NAME: string = 'm-progress';
export const RADIO_NAME: string = 'm-radio';
export const RADIO_GROUP_NAME: string = 'm-radio-group';
export const RADIO_STYLE_NAME: string = 'm-radio-style';
export const SCROLL_TOP_NAME: string = 'm-scroll-top';
export const SIDEBAR_NAME: string = 'm-sidebar';
export const SLIDER_NAME: string = 'm-slider';
export const SPINNER_NAME: string = 'm-spinner';
export const STATUS_NAME: string = 'm-status';
export const STEP_NAME: string = 'm-step';
export const STEPPERS_NAME: string = 'm-steppers';
export const STEPPERS_ITEM_NAME: string = 'm-steppers-item';
export const SWITCH_NAME: string = 'm-switch';
export const TAB_PANEL_NAME: string = 'm-tab-panel';
export const TABS_NAME: string = 'm-tabs';
export const TEMPLATE_NAME: string = 'm-template';
export const TEXTAREA_NAME: string = 'm-textarea';
export const TEXTFIELD_NAME: string = 'm-textfield';
export const TIMEPICKER_NAME: string = 'm-timepicker';
export const TOOLTIP_NAME: string = 'm-tooltip';
export const UPLOAD_NAME: string = 'm-upload';
export const UPLOAD_INPUT_NAME: string = 'm-upload-input';
export const UPLOAD_DRAGDROP_NAME: string = 'm-upload-dragdrop';
export const UPLOAD_FILESLIST_NAME: string = 'm-upload-fileslist';
export const VALIDATION_MESSAGE_NAME: string = 'm-validation-message';
