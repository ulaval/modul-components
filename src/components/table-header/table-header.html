<div class="m-table-header">
    <slot name="left"></slot>
    <div class="m-table-header__spacer"></div>
    <slot name="right"></slot>
    <m-link class="m-table-header__add-btn"
            :class="{'m--has-margin': $slots.right}"
            v-if="addBtnLabel"
            :icon="true"
            icon-name="m-svg__add-circle-filled"
            icon-size="22px"
            mode="button"
            @click="add"
            ref="addBtn">{{ addBtnLabel }}</m-link>
</div>
