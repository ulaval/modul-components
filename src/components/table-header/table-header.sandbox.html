<div>
    <h2>Default</h2>
    <m-table-header></m-table-header>

    <h2>With addBtnLabel</h2>
    <m-table-header add-btn-label="Add data"></m-table-header>

    <h2>With left slot</h2>
    <m-table-header>
        <div slot="left">Content in left slot</div>
    </m-table-header>

    <h2>With right slot</h2>
    <m-table-header>
        <div slot="right">Content in right slot</div>
    </m-table-header>

    <h2>With addBtnLabel, left and right slots</h2>
    <m-table-header add-btn-label="Add data">
        <div slot="left">Content in left slot</div>
        <div slot="right">Content in right slot</div>
    </m-table-header>
</div>
