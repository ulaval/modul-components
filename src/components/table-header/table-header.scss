@import 'commons';

.m-table-header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: $m-spacing;

    &__spacer {
        flex: 1 1 auto;
        min-width: $m-spacing;
    }

    &__add-btn.m--has-margin {
        margin-left: $m-spacing;
    }
}
