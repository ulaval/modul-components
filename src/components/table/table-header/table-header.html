<thead>
    <m-table-row>
        <slot></slot>
    </m-table-row>
</thead>
