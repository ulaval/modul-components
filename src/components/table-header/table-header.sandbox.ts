import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { TABLE_HEADER_NAME } from '../component-names';
import WithRender from './table-header.sandbox.html';
import TableHeaderPlugin from './table-header';

@WithRender
@Component
export class MTableHeaderSandbox extends Vue {
}

const TableHeaderSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(TableHeaderPlugin);
        v.component(`${TABLE_HEADER_NAME}-sandbox`, MTableHeaderSandbox);
    }
};

export default TableHeaderSandboxPlugin;
