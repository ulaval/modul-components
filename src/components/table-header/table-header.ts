import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Emit, Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
import { TABLE_HEADER_NAME } from '../component-names';
import { MIconButtonSkin } from '../icon-button/icon-button';
import WithRender from './table-header.html?style=./table-header.scss';
import LinkPlugin from '../link/link';

@WithRender
@Component
export class MTableHeader extends ModulVue {
    @Prop()
    addBtnLabel: string;

    addButtonSkin: MIconButtonSkin = MIconButtonSkin.Link;

    @Emit('add')
    add(): void {
    }
}

const TableHeaderPlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(LinkPlugin);
        v.component(TABLE_HEADER_NAME, MTableHeader);
    }
};

export default TableHeaderPlugin;
