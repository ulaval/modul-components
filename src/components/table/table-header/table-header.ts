import { PluginObject } from 'vue';
import Component from 'vue-class-component';

import { ModulVue } from '../../../utils/vue/vue';
import { TABLE_HEADER_NAME } from '../../component-names';
import WithRender from './table-header.html';

@WithRender
@Component
export class MTableHeader extends ModulVue {

}

const TableHeaderPlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(TABLE_HEADER_NAME, 'plugin.install');
        v.component(TABLE_HEADER_NAME, MTableHeader);
    }
};

export default TableHeaderPlugin;
