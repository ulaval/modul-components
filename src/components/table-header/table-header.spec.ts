import { RefSelector, shallowMount, Wrapper } from '@vue/test-utils';
import { renderComponent } from '../../../tests/helpers/render';
import { TABLE_HEADER_NAME } from '../component-names';
import { MTableHeader } from './table-header';

const ADD_BTN_LABEL: string = 'ADD BUTTON';

const REF_ADD_BTN: RefSelector = { ref: 'addBtn' };

let slots: any = {};
const SLOT_LEFT: string = '<div>SLOT LEFT</div>';
const SLOT_RIGHT: string = '<div>SLOT RIGHT</div>';

let wrapper: Wrapper<MTableHeader>;

const initializeShallowWrapper: any = () => {
    wrapper = shallowMount(MTableHeader, {
        slots
    });
};

describe(TABLE_HEADER_NAME, () => {
    beforeEach(() => {
        initializeShallowWrapper();
    });

    it(`Should render correctly`, () => {
        expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
    });

    it(`Should have an empty add button label`, () => {
        expect(wrapper.props().addBtnLabel).toBeUndefined();
    });

    describe(`Given a add button label`, () => {
        beforeEach(() => {
            initializeShallowWrapper();
            wrapper.setProps({ addBtnLabel: ADD_BTN_LABEL });
        });

        it(`Should render add button with label from prop`, () => {
            expect(wrapper.find(REF_ADD_BTN).text()).toBe(ADD_BTN_LABEL);
        });

        it(`Should emit "add" event when clicked`, () => {
            wrapper.find(REF_ADD_BTN).trigger('click');

            expect(wrapper.emitted('add')).toBeTruthy();
        });
    });

    describe(`Given content in left slot`, () => {
        it(`Should render correctly`, () => {
            slots = {
                left: SLOT_LEFT
            };
            initializeShallowWrapper();

            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });
    });

    describe(`Given content in right slot`, () => {
        it(`Should render correctly`, () => {
            slots = {
                right: SLOT_RIGHT
            };
            initializeShallowWrapper();

            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });
    });

    describe(`Given a add button label, left and right slots`, () => {
        it(`Should render correctly`, () => {
            slots = {
                left: SLOT_LEFT,
                right: SLOT_RIGHT
            };
            initializeShallowWrapper();
            wrapper.setProps({ addBtnLabel: ADD_BTN_LABEL });

            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });
    });
});
