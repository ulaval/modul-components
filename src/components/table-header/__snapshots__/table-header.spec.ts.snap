// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`m-table-header Given a add button label, left and right slots Should render correctly 1`] = `
<div class="m-table-header">
    <div>SLOT LEFT</div>
    <div class="m-table-header__spacer"></div>
    <div>SLOT RIGHT</div>
    <m-link icon="true" icon-name="m-svg__add-circle-filled" icon-size="22px" mode="button" class="m-table-header__add-btn m--has-margin">ADD BUTTON</m-link>
</div>
`;

exports[`m-table-header Given content in left slot Should render correctly 1`] = `
<div class="m-table-header">
    <div>SLOT LEFT</div>
    <div class="m-table-header__spacer"></div>
</div>
`;

exports[`m-table-header Given content in right slot Should render correctly 1`] = `
<div class="m-table-header">
    <div class="m-table-header__spacer"></div>
    <div>SLOT RIGHT</div>
</div>
`;

exports[`m-table-header Should render correctly 1`] = `
<div class="m-table-header">
    <div class="m-table-header__spacer"></div>
</div>
`;
